### PR TITLE
feat: support refs in rulesets

### PR DESCRIPTION
**Pull request by P0lip**
Closes https://github.com/stoplightio/spectral/issues/380

**Checklist**

- [x] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

NOTE: I'm not planning to add any docs - I don't think we should advertise this feature.

### CODE REVIEW

**Comment by P0lip** [comment]
@marbemac @philsturgeon 
I got it working yet the tests fail when executed in parallel.
This is caused by some json-ref-resolver quirky behavior.
Will verify why this happens.

That said, I'd rather revert that change and go with a simpler concept. It does increase complexity in terms of Studio and I don't' really see whether $refs in rulesets would be that useful given `extends` we have.

**Comment by marbemac** [comment]
What’s the alternative that leverages extends to accomplish the goal?

Re this implementation, rather than making load ruleset async, what about leveraging the async run (I think it’s called run? Not sure, on phone, but something like this)  method we already have? Basically at the start of run, resolve the merged ruleset if it has not been previously resolved. Could hash the merged ruleset and use that as a cache key so that multiple invocations of run only need to resolve once, unless the underlying ruleset object has changed.

**Comment by philsturgeon** [comment]
Just to remind everyone, the original use case was "we want to get the inline JSON Schema definitions out of the ruleset files and into their own locations" so if this larger more generic feature is causing trouble, maybe we can take a step back, and look at just solving the original use case? Sometimes I feel like making things more general is more trouble than its worth.

**Comment by P0lip** [comment]
> Just to remind everyone, the original use case was "we want to get the inline JSON Schema definitions out of the ruleset files and into their own files" so if this larger more generic feature is causing trouble, maybe we can take a step back, and look at just solving the original use case? Sometimes I feel like making things more general is more trouble than its worth.

Yeah, exactly. I mean, we are bringing something very robust that is by far more flexible than we need it to be. It's certainly something we'll need to maintain even if we keep it undocumented, as others might just discover that and use it in their own rulesets.
Either way, I'm pretty much there - it works like a charm when tests are executed in full isolation (one by one), when they are run in parallel they fail for some reason.

**Comment by P0lip** [comment]
> What’s the alternative that leverages extends to accomplish the goal?

`extends: [oas2-schema]`?
I know it might look awkward, but it'd its job and end developers wouldn't even notice since they extend `oas2` or `oas3` rulesets.

>  Not sure, on phone, but something like this) method we already have? Basically at the start of run, resolve the merged ruleset if it has not been previously resolved. 

This would work for `schema`, but what if one places `$ref` in a rule? In order to merge everything safely, we'd need to operate on resolved data.

**Comment by P0lip** [comment]
@philsturgeon @marbemac 
Okay, I think I nailed the issue.
It seems like `expect` is having some serious troubles with our assertions.
Not sure if it's diffing or just the result printing, but basically if I adjust the assertion a bit, it all works fine.

**Comment by P0lip** [comment]
https://github.com/stoplightio/spectral/pull/454/commits/c4d4365e8a8603aa36eb58393df0860784ae128e is unrelated to the PR, but it's a fix for an issue exposed by the PR - we didn't handle schema ids properly.

File: src/rulesets/oas2/__tests__/oas2-schema.ts [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
   const s = new Spectral();
+
   s.addRules({
-    'oas2-schema': Object.assign(ruleset.rules['oas2-schema'], {
+    'oas2-schema': Object.assign({}, ruleset.rules['oas2-schema'], {
       recommended: true,
       type: RuleType[ruleset.rules['oas2-schema'].type],
+      then: {
+        ...ruleset.rules['oas2-schema'].then,
+        functionOptions: {
```
Comment (philsturgeon) [inline src/rulesets/oas2/__tests__/oas2-schema.ts]: Is this meant to be here? Looks like we’re oassing in the contents not a ref. Is that just to make the unit test work and avoid it being an integration test? If so, should we do the same for oas3?

Comment (P0lip) [reply src/rulesets/oas2/__tests__/oas2-schema.ts]: Yeah, it's to avoid it being an integration test.
We don't have any tests for oas3-schema

**Comment by philsturgeon** [comment]
YAAAAY

**Pull request merged**